kv/range_cache.go

```diff
@@ -80,13 +80,75 @@ type rangeDescriptorCache struct {
 	rangeCache *cache.OrderedCache
 	// rangeCacheMu protects rangeCache for concurrent access
 	rangeCacheMu sync.RWMutex
+	// lookupRequests stores all inflight requests retrieving range
+	// descriptors from the database. It allows multiple rangeDescriptorCache
+	// lookup requests for the same (key, considerIntents, useReverseScan)
+	// tuple to be multiplexed onto the same database lookup.
+	lookupRequests struct {
+		sync.Mutex
+		inflight map[lookupRequestKey]lookupRequest
+	}
+	// splitPrefetchSync sets whether prefetching the updated half of a split
+	// when a split is detected should be synchronous or completed in the background.
+	splitPrefetchSync bool
+}
+
+type lookupRequest struct {
+	// observers stores the cache lookups that have joined onto the existing
+	// lookupRequest and wish to be notified of its database lookup's results
+	// instead of performing the database lookup themselves.
+	observers []chan<- lookupResult
+}
+
+type lookupResult struct {
+	desc *roachpb.RangeDescriptor
+	pErr *roachpb.Error
+}
+
+type lookupRequestKey struct {
+	key             string
+	evictedPrev     bool
```

Comment:
Right now we use `evicted` state directly to assume a key missed on the right half of a split. I'm wondering if we should focus this flag so that it's only making the assumption for certain kind of errors (like `RangeKeyMismatchError`). That way, an eviction due to a `NotLeaderError` or `SendError` wont get grouped in with this right half assumption request.